handlejoin.go

```diff
@@ -0,0 +1,278 @@
+package gomatrixserverlib
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/matrix-org/gomatrixserverlib/spec"
+	"github.com/matrix-org/util"
+)
+
+type HandleMakeJoinInput struct {
+	Context            context.Context
+	UserID             spec.UserID
+	RoomID             spec.RoomID
+	RoomVersion        RoomVersion
+	RemoteVersions     []RoomVersion
+	RequestOrigin      spec.ServerName
+	RequestDestination spec.ServerName
+	LocalServerName    spec.ServerName
+	LocalServerInRoom  bool
+	RoomQuerier        JoinRoomQuerier
+	BuildEventTemplate func(*ProtoEvent) (PDU, []PDU, error)
+}
+
+type HandleMakeJoinResponse struct {
+	JoinTemplateEvent ProtoEvent
+	RoomVersion       RoomVersion
+}
+
+func HandleMakeJoin(input HandleMakeJoinInput) (*HandleMakeJoinResponse, error) {
```

Comment:
Naming: We're using `HandleMakeJoin` everywhere (this function, Response, and Input). Why not just `JoinRoom`? This is a top-level function, so shouldn't be confused with the `fclient` version?

Reply:
Because this is the receiving side